reark/src/main/java/io/reark/reark/data/store/ContentProviderStore.java

```diff
@@ -57,66 +64,124 @@
 
     protected ContentProviderStore(@NonNull ContentResolver contentResolver) {
+        this(contentResolver, 100, 30);
+    }
+
+    protected ContentProviderStore(@NonNull ContentResolver contentResolver, int groupingTimeout, int groupMaxSize) {
         Preconditions.checkNotNull(contentResolver, "Content Resolver cannot be null.");
 
         this.contentResolver = contentResolver;
         this.contentResolver.registerContentObserver(getContentUri(), true, contentObserver);
 
-        updateSubject
-                .onBackpressureBuffer()
-                .observeOn(Schedulers.io())
-                .subscribe(pair -> {
-                    updateIfValueChanged(this, pair);
-                });
+        this.groupingTimeout = groupingTimeout;
+        this.groupMaxSize = groupMaxSize;
+
+        // Observable transforming updates and inserts to ContentProviderOperations
+        Observable<ContentProviderOperation> operationObservable = updateSubject
+                .flatMap(pair -> createOperation(pair.first, pair.second))
+                .filter(new NullFilter())
+                .observeOn(Schedulers.computation());
+
+        // Group the operations to a list that should be executed in one batch. The default
+        // grouping logic is suitable for pojo stores, but some stores may need to provide
+        // their own grouping logic, if for example buffering delays are undesirable.
+        groupOperations(operationObservable)
+                .doOnNext(list -> Log.v(TAG, "Grouped list of " + list.size()))
+                .doOnNext(operations -> {
+                    try {
+                        ContentProviderResult[] result = contentResolver.applyBatch(getAuthority(), operations);
+                        Log.v(TAG, String.format("Applied %s operations", result.length));
+                    } catch (RemoteException | OperationApplicationException e) {
+                        Log.e(TAG, "Error applying operations", e);
+                    }
+                })
+                .flatMap(Observable::from)
+                .map(ContentProviderOperation::getUri)
+                .subscribe(locker::release, Log.onError(TAG));
     }
 
-    private static <T> void updateIfValueChanged(ContentProviderStore<T> store, Pair<T, Uri> pair) {
-        final Cursor cursor = store.contentResolver.query(pair.second, store.getProjection(), null, null, null);
-        T newItem = pair.first;
-        boolean valuesEqual = false;
+    /**
+     * Implements grouping logic for batching the content provider operations. The default
+     * logic buffers the operations with debounced timer, while applying a hard limit for the
+     * number of operations. The data is serialized into a binder transaction, and an attempt
+     * to pass a too large batch of operations will result in a failed binder transaction.
+     */
+    protected Observable<ArrayList<ContentProviderOperation>> groupOperations(Observable<ContentProviderOperation> source) {
```

Comment:
Yes, for whatever reason contentResolver.applyBatch() takes ArrayList. Changed the List -> ArrayList mapping to happen just before it's needed though.